### PR TITLE
[Merged by Bors] - feat(GroupTheory/Frattini): definition and first theorem

### PR DESCRIPTION
**Pull request by colva**
We give the definition of the Frattini subgroup of a group, and three elementary results:
* The Frattini subgroup is characteristic.
* If every subgroup of a group is contained in a maximal subgroup, then
the Frattini subgroup consists of the non-generating elements of the group.
* The Frattini subgroup of a finite group is nilpotent.

Co-authored-by: Inna Capdeboscq <inna.capdeboscq@orange.fr>
Co-authored-by: Susanna Fishel <sfishel1@asu.edu>
Co-authored-by: Kim Morrison <kim@tqft.net>

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Review by tb65536 — COMMENTED** [review]
Some of your API can be generalized from bijections to surjections. For example:
```
namespace Subgroup

variable {G H : Type*} [Group G] [Group H] {φ : G →* H} (hφ : Function.Surjective φ)

lemma coatom_comap_of_surjective {M : Subgroup H} (hM : IsCoatom M)  :
    IsCoatom (M.comap φ) := by
  refine And.imp (fun hM ↦ ?_) (fun hM ↦ ?_) hM
  · rwa [← (comap_injective hφ).ne_iff, comap_top] at hM
  · intro K hK
    specialize hM (K.map φ)
    rw [← comap_lt_comap_of_surjective hφ, ← (comap_injective hφ).eq_iff] at hM
    rw [comap_map_eq_self ((M.ker_le_comap φ).trans hK.le), comap_top] at hM
    exact hM hK

lemma frattini_le_comap_frattini_of_surjective : frattini G ≤ (frattini H).comap φ := by
  simp_rw [frattini, Order.radical, comap_iInf, le_iInf_iff]
  intro M hM
  apply biInf_le
  exact coatom_comap_of_surjective hφ hM

lemma map_frattini_le_frattini_of_surjective : (frattini G).map φ ≤ frattini H := by
  rw [map_le_iff_le_comap]
  exact frattini_le_comap_frattini_of_surjective hφ

/-- The Frattini subgroup is characteristic. -/
instance frattini_characteristic : (frattini G).Characteristic := by
  rw [characteristic_iff_map_le]
  intro φ
  exact map_frattini_le_frattini_of_surjective φ.surjective

end Subgroup
```

**Comment by semorrison** [comment]
Thanks, @tb65536, we've added this. I didn't immediately see a lattice theoretic version of your statement.

File: Mathlib/GroupTheory/Frattini.lean [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+If a subgroup together with the Frattini subgroup generates the whole group,
+then the subgroup is already the whole group.
+-/
+theorem frattini_nongenerating [IsCoatomic (Subgroup G)] {K : Subgroup G}
+    (h : K ⊔ frattini G = ⊤) : K = ⊤ :=
+  Order.radical_nongenerating h
+
+-- The Sylow files unnecessarily use `Fintype` (computable) where often `Finite` would suffice,
+-- so wwe need this:
```
Comment (tb65536) [inline Mathlib/GroupTheory/Frattini.lean:55]: (Also, I'm going to try to clean up this fintype/finite mess in the group theory library, starting with #13431 and #13432, but eventually getting to p-groups and Sylow).
```suggestion
-- so we need this:
```

**Review by tb65536 — APPROVED** [review]
LGTM

**Review by rosborn — APPROVED** [review]
Looks good!

**Comment by semorrison** [comment]
(btw, no need to set something as `awaiting-author` just to correct a typo like `wwe` --> `we` --- just hit commit yourself! :-)

**Comment by tb65536** [comment]
Ah, good point.

maintainer merge

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**